frontend/src/lib/components/ActivityLog/activityLogLogic.tsx

```diff
@@ -1,30 +1,100 @@
     props: {} as ActivityLogProps,
     key: ({ scope, id }) => `activity/${scope}/${id || 'all'}`,
-    loaders: ({ props }) => ({
-        activity: [
-            [] as HumanizedActivityLogItem[],
+    loaders: ({ values }) => ({
+        nextPage: [
+            { results: [] as ActivityLogItem[], total_count: 0 } as CountedPaginatedResponse,
+            {
+                fetchNextPage: async () => {
+                    const url = values.nextPageURL
+                    return url === null ? null : await api.get(url)
+                },
+            },
+        ],
+        previousPage: [
+            { results: [] as ActivityLogItem[], total_count: 0 } as CountedPaginatedResponse,
             {
-                fetchActivity: async () => {
-                    const url = props.id
-                        ? `/api/projects/@current/feature_flags/${props.id}/activity`
-                        : `/api/projects/@current/feature_flags/activity`
-                    const apiResponse: PaginatedResponse<ActivityLogItem> = await api.get(url)
-                    return humanize(apiResponse?.results, props.describer)
+                fetchPreviousPage: async () => {
+                    const url = values.previousPageURL
+                    return url === null ? null : await api.get(url)
                 },
             },
         ],
     }),
+    reducers: ({ props }) => ({
+        page: [
+            props.startingPage ? props.startingPage - 1 : 0,
+            {
+                fetchNextPageSuccess: (state) => state + 1,
+                fetchPreviousPageSuccess: (state) => state - 1,
+            },
+        ],
+        humanizedActivity: [
+            [] as HumanizedActivityLogItem[],
+            {
+                fetchNextPageSuccess: (state, { nextPage }) =>
+                    nextPage ? humanize(nextPage.results, props.describer) : state,
+                fetchPreviousPageSuccess: (state, { previousPage }) =>
+                    previousPage ? humanize(previousPage.results, props.describer) : state,
+            },
+        ],
+        previousPageURL: [
+            null as string | null,
+            {
+                fetchNextPageSuccess: (_, { nextPage }) => nextPage.previous || null,
+                fetchPreviousPageSuccess: (_, { previousPage }) => previousPage.previous || null,
+            },
+        ],
+        totalCount: [
+            null as number | null,
+            {
+                fetchNextPageSuccess: (_, { nextPage }) => nextPage.total_count || null,
+                fetchPreviousPageSuccess: (_, { previousPage }) => previousPage.total_count || null,
+            },
+        ],
+        nextPageURL: [
+            (props.id
+                ? `/api/projects/@current/feature_flags/${props.id}/activity?page=${props.startingPage || 1}`
+                : `/api/projects/@current/feature_flags/activity?page=${props.startingPage || 1}`) as string | null,
```

Comment:
Could these somehow go through the api builder methods in `api.ts`?

Reply:
Hmm, these move in PR #9226 

I'll make the change to `api.ts` there. To avoid unpicking it at merge-time